Adjust request lsn at ps

knizhnik opened the pull request:
## Problem

See  https://databricks.slack.com/archives/C09254R641L/p1752245110754979

## Summary of changes

1. Set request_lsn=max(request_lsn, gc_cutoff_lsn) at page server
2. Check page LSN in replica and report error if it is larger than apply_lsn (which can only happen in case of expired lease)

problame commented on pgxn/neon/communicator.c:
```diff
@@ -2227,6 +2227,23 @@ communicator_read_at_lsnv(NRelFileInfo rinfo, ForkNumber forkNum, BlockNumber ba
+#if PG_VERSION_NUM >= 150000
+					XLogRecPtr replay_lsn = GetCurrentReplayRecPtr(NULL);
+#else
+					XLogRecPtr replay_lsn = Max(GetXLogReplayRecPtr(NULL), resp->lsn);
+#endif
```
Why is it correct to consider `resp->lsn` as the `replay_lsn` here?

problame commented on pageserver/src/page_service.rs:
```diff
@@ -2195,12 +2195,16 @@ impl PageServerHandler {
+                    "Tried to request a page version that was garbage collected. requested at {} gc cutoff {}",
+                    request_lsn, **latest_gc_cutoff_lsn
                 );
+                request_lsn = **latest_gc_cutoff_lsn;
```
What if the page doesn't exist anymore at this raised LSN e.g. because we truncate the relation?
It will cause scary-looking errors that look like data loss to pageserver operators until they spot that we artificially raised this request_lsn.

(Pageserver team operates under the assumption that reads from compute are legitimate, i.e., point to a key that can be read at the effective request lsn; we currently treat read failures as serious errors that warrant immediate investigation)

knizhnik replied:
I agree - it can be the problem.
Compute request some page at Lsn1, then relation is truncated at Lsn2>Lsn1 and current redo position at replica is Lsn3>Lsn2.

But please notice that this issue is not somehow related to the problem with leases, standaby_horizon and even replicas.
It can happen even without any GC at PS side. Compute send some prefetch requests for page P1. And then this query is completed, locks are released and some other backend truncates or drops the relation. 

In this case when PS receives this prefetch request, this page doesn't exists. In this case PS has to return error.
This errors will; be received by compute while consuming prefetch requests. But it will nto be propagated to the client because this page is not read.

In case of replicas and problems with leases (please notice only in this abnormal case!) this error can really propagated to the client. But this is why we print this warning in PS log. So if case if investigation of possible corruption issue, we will see this warning and understand that most likely it is caused by lease expiration.

problame commented on pageserver/src/page_service.rs:
```diff
@@ -2195,12 +2195,16 @@ impl PageServerHandler {
+                // latest_gc_cutoff_lsn is conservative estimation for min(redo_lsn) for all replicas,
+                // so it is safe to move request_lsn forward to latest_gc_cutoff_lsn.
+                // If replica lease is expired and latest_gc_cutoff_lsn>redo_lsn for this replica,
+                // then check of page LSN at replia protects it from getting too new version of the page.
```
This PR only adds the check for RO replicas in the communicator, but, this method here is also executed for requests from primary and static computes.
Practically, 
- primary sends request_lsn=0xfff..., so, never going to hit the outer `if` conditino, but, that's not obvious to future readers of this code and
- static compute sends request_lsn=$static_lsn , which, may very well hit this branch here and raise the request_lsn

I guess in contrast to commentary in communicator.c, the check there also applies to static computes?
But, not super obvious, at very least the comment there requires adjustment.

knizhnik replied:
If LSN requested by static replica is collected by GC and we have to advance `request_lsn` to `gc_cutoff` here, then it will be detected by check in compute that returned page LSN is greater than `redo_lsn=static_lsn`.

knizhnik commented on pgxn/neon/communicator.c:
```diff
@@ -2227,6 +2227,23 @@ communicator_read_at_lsnv(NRelFileInfo rinfo, ForkNumber forkNum, BlockNumber ba
 			case T_NeonGetPageResponse:
 			{
 				NeonGetPageResponse* getpage_resp = (NeonGetPageResponse *) resp;
+				if (resp->lsn != UINT64_MAX) /* replica */
```
1. There is separate check that requests matches response.
2. It is the simplest one:) "Canonical" way is to call `RecoveryInProgress()`. But it can not be used at startup when shared memory is not yet initialized. can SMGR be accessed at startup? Most likely - not. But I am not 100% sure.

knizhnik commented:
> How does this PR handle protocol v2 responses? Does it need to?

This check doesn't require request data available only in v3 protocol version, except this hack for pg14.
I consider this combination of v2+pg14 as something very exotic (we already switched to v3 everywhere) to make some special effort here.

> prefetch_wait_for appears to already to a bunch of checks, including calling neon_prefetch_response_usable() which seems like the more appropriate / centralized place for a check like this one?

No, it is not called on all paths now.

> I think there's a bug where prefetch_wait_for->prefetch_read->lfc_prefetch call fills in the page without performing the check you added?

Yes, thank you.
Fixed.

> As a maintainer/on-call person for Pageserver, I'm unconfortable with the implictness of how we identify replica OR static_compute versus primary, based solely on the very implicit knowledge that primary sends request_lsn = 0xfff...

Changed to `RecoveryInPorgress()`

> Generally, I'm concerned with lack of documenttion aronud the protocol semantics for pagestream v3 protocol and/or comunicator protocol.
It should be documented somewhere, in written English, that with this PR, getpage responses now will return response header lsn field as in the request but
pd_lsn in the page image can now be > that lsn, and the request's not_modified_since LSN for that matter.

+1

hlinnaka commented:
Please explain what the original problem is in the PR description. A link to a slack thread will be useless after some time.

ololobus commented:
> 1. Set request_lsn=max(request_lsn, gc_cutoff_lsn) at page server

afaik, we decided to do not do that, right? @knizhnik

knizhnik commented:
We defined to implement another approach: https://github.com/neondatabase/neon/pull/12634
